CertAuthority watcher filtering

espadolini opened the pull request:
This PR:
- fixes the filtering on CA updates that trigger a rotation state sync
- makes it so that watchers can filter `CertAuthority` resources by type (`host`/`user`/`jwt`) and name, and adds the appropriate CA filter to the cache for `Node` roles (host CAs with a name that matches the cluster name, all user CAs) and for the `syncRotationStateCycle` loop (only host CAs from the current cluster)
- adds the same filtering support to the cache initialization, so we don't fetch unnecessary data
- adds filtering on the server side for older node clients, gated by version number
- fixes potential inconsistencies in CAs when adding a trusted cluster/remote cluster relationship
- filters out spurious CA updates that would result in no changes

Fixes #9895.

The version cutoff for applying the server-side filter was set to 9.0.0; any backport at v7 and earlier should remove the client side filters and rely on that, as clients are allowed to connect to auth servers with a higher major version that don't have the backport yet.

espadolini committed:
Fix potential inconsistency in CAs when adding a remote cluster

espadolini committed:
Fix filtering on CA updates that should trigger a rotation state sync

espadolini committed:
Keep track of the trust relationship of CertAuthorities

espadolini committed:
Allow filtering CertAuthority watchers by trust relationship

espadolini committed:
WIP: CertAuthority filtering on the server side for older nodes

espadolini commented:
Open questions:
- [ ] should we add a `CertAuthorityFilter` in protobuf form, if it's only ever used for translating to and from a watcher's filter map?
- [ ] does the design of the filter need product decision? It's technically exposed API
- [ ] where should the backwards compatibility filter enforcer live? `auth_with_roles`? `grpcserver`? The only other version check on a client is in `grpcserver` but checking roles in there seems a bit of a layering violation.
- [ ] how should the version cutoff work? Anything less than the next point release at time of backport? Should it be backport-only?

fspmarshall reviewed (dismissed):
Added some minor nits, but LGTM overall.

espadolini committed:
Add a CertAuthorityFilter utility type

espadolini committed:
Renames to make trust relationships more clear

espadolini committed:
Version cutoff for the server-side filter injection

rosstimothy reviewed (commented):
Overall looks good. Can you add some test coverage for the things you added/fixed?

espadolini committed:
Add trust relationship checks in the trustedClusters() integration test

espadolini committed:
Clean up integration/trustedClusters

espadolini committed:
Add CA trust relationships to lib/cache:TestWatchers

espadolini committed:
Add a test for the CA filtering in the ForNode cache configuration

espadolini committed:
Tweak CertAuthorityFilter

espadolini committed:
Improve CA trust labeling

Includes improvements in the validation of CAs received when adding
a trusted cluster.

espadolini committed:
Outdated code removal

espadolini committed:
Don't let CAs with unspecified trust through filters that block everything.

espadolini committed:
Store "local" trust rels as no label

This lets us upgrade and downgrade with relative impunity, as trust
relationships won't ever change without deleting and recreating
a trusted cluster/remote cluster, and we only ever store nonlocal
CAs that are local from the point of view of the other end, and thus
will have no label when fetched in any way

espadolini committed:
Apply the watch filter in the initial fetch in the certAuthority collector

rosstimothy commented on integration/integration_test.go:
```diff
@@ -2122,15 +2123,52 @@ func trustedClusters(t *testing.T, suite *integrationTestSuite, test trustedClus
+		name     string
+		trustRel types.TrustRelationship // empty if we expect the CA to not exist
+	}{
+		{main, types.HostCA, clusterMain, types.TrustRelationshipLocal},
```
nit: use named initialization
```suggestion
		{instance: main, caType: types.HostCA, name: clusterMain, trustRel: types.TrustRelationshipLocal},
```

espadolini committed:
Skip upserting equivalent CAs in RotateExternalCertAuthority

espadolini committed:
Switch the CA filtering to name+type instead

This still lets us do the CA filtering for the Node cache and the
syncRotationStateCycle watcher without adding new data to CAs.
As a bonus, the cache can now fetch only the required CAs on
initialization.

espadolini committed:
Ignore unknown WatchKind filters for the future

espadolini committed:
Remove trust labels

espadolini committed:
Don't unnecessarily upsert unchanged CAs

espadolini commented on api/types/authority.go:
```diff
@@ -723,3 +782,63 @@ func (k *JWTKeyPair) CheckAndSetDefaults() error {
+		return nil
+	}
+
+	*f = make(CertAuthorityFilter, len(m))
```
The conversion is now infallible, so nothing can go wrong 😅

rosstimothy commented on lib/auth/grpcserver.go:
```diff
@@ -334,6 +340,56 @@ func (g *GRPCServer) WatchEvents(watch *proto.Watch, stream proto.AuthService_Wa
+
+// certAuthorityFilterVersionCutoff is the version starting from which we stop
+// injecting filters for CertAuthority watches in maybeFilterCertAuthorityWatches.
+var certAuthorityFilterVersionCutoff = *semver.New("8.1.3")
```
Is this still the correct version? 8.3.0 came out yesterday.

espadolini replied:
No but thank you for the reminder; I'll add a note to the PR description stating that the version needs to be updated to the latest release before merging this. Not sure if there's a better way of doing that.

espadolini commented on lib/auth/grpcserver.go:
```diff
@@ -334,6 +340,56 @@ func (g *GRPCServer) WatchEvents(watch *proto.Watch, stream proto.AuthService_Wa
+		log.Debugf("Injecting filter for CertAuthority watch for Node-only watcher with version %v", clientVersion)
+		watch.Kinds[i].Filter = types.CertAuthorityFilter{
+			types.HostCA: clusterName,
+			types.UserCA: "*",
```
Will do

espadolini commented on lib/cache/collections.go:
```diff
@@ -809,56 +812,91 @@ func (c *certAuthority) erase(ctx context.Context) error {
 }
 
 func (c *certAuthority) fetch(ctx context.Context) (apply func(ctx context.Context) error, err error) {
-	applyHostCAs, err := c.fetchCertAuthorities(types.HostCA)
-	if err != nil {
-		return nil, trace.Wrap(err)
+	applyFns := []func(ctx context.Context) error{}
+
+	for _, caType := range types.CertAuthTypes {
+		var fn func(ctx context.Context) error
+		var err error
+		if c.filter.IsEmpty() || c.filter[caType] == "*" {
+			fn, err = c.fetchCertAuthorities(caType)
+		} else if c.filter[caType] != "" {
+			fn, err = c.fetchCertAuthority(types.CertAuthID{
+				Type:       caType,
+				DomainName: c.filter[caType],
+			})
+		} else {
+			fn = c.deleteAllCertAuthorities(caType)
+		}
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		applyFns = append(applyFns, fn)
```
This is in a separate commit on purpose, want me to revert it so the implementation is simpler? We would lose the benefit of not fetching all the CAs on init, and there might be more than a few of those - although maybe not enough to matter, in the grand scheme of things?

espadolini replied:
Reverted the changes to the cache init, the `certAuthority` collector now fetches as before and then filters out the returned CAs with the regular `Match` method.

espadolini committed:
Revert changes to the `certAuthority` collector

espadolini committed:
Undo spurious changes that are no longer related

espadolini committed:
Use Wildcard instead of a literal asterisk in CertAuthorityFilter

espadolini commented on lib/auth/grpcserver.go:
```diff
@@ -334,6 +340,56 @@ func (g *GRPCServer) WatchEvents(watch *proto.Watch, stream proto.AuthService_Wa
+//
+// DELETE IN 10.0, no supported clients should require this at that point
+func maybeFilterCertAuthorityWatches(ctx context.Context, clusterName string, roleNames []string, watch *types.Watch) {
+	if len(roleNames) != 1 || roleNames[0] != string(types.RoleNode) {
+		return
+	}
```
Check's in place

espadolini committed:
Bump the version cutoff for the filter injection to v9

r0mant commented on lib/services/local/trust.go:
```diff
@@ -73,6 +73,17 @@ func (s *CA) UpsertCertAuthority(ca types.CertAuthority) error {
+		Type:       ca.GetType(),
+		DomainName: ca.GetClusterName(),
+	}, true); err == nil {
+		if services.CertAuthoritiesEquivalent(existing, ca) {
+			return nil
+		}
```
@espadolini @fspmarshall I'm a little worried that silently skipping updates like this might result into very subtle bugs where e.g. we forget to update CertAuthoritiesEquivalent method after introducing another CA field or something. What do you think about at least logging it here?

espadolini replied:
`CertAuthoritiesEquivalent` is defined as cmp.Equal skipping the metadata ID field so we shouldn't miss anything. Adding a debug log line would be fine, anything more would get quite spammy however.

r0mant replied:
Ah yes, I forgot it does cmp.Equal, should be fine then.

r0mant commented on lib/service/connect.go:
```diff
@@ -565,7 +570,7 @@ func (process *TeleportProcess) syncRotationStateCycle() error {
 				continue
 			}
-			if ca.GetType() != types.HostCA && ca.GetClusterName() != conn.ClientIdentity.ClusterName {
+			if ca.GetType() != types.HostCA || ca.GetClusterName() != conn.ClientIdentity.ClusterName {
```
Is that intended change?

espadolini replied:
Yes, the original version had a bug and called syncRotationStateAndBroadcast unnecessarily often; syncRotationStateAndBroadcast only ever cares about the local host CA, so it should be called when just that one has potentially changed.

r0mant replied:
Got it, thanks for the clarification.

r0mant committed:
Merge branch 'master' into espadolini/ca-watcher-filter
